File: Typecast/TCGlyphWindowController.m
//
//  TCGlyphWindowController.m
//  Type Designer
//
//  Created by David Schweinsberg on 19/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCGlyphWindowController.h"
#import "TCDocument.h"
#import "TCGlyph.h"
#import "TCGlyphView.h"
#import "TCHeadTable.h"
#import "TCHheaTable.h"
#import "TCHmtxTable.h"
#import "TCFontCollection.h"
#import "TCFont.h"

@interface TCGlyphWindowController ()

@property (weak) IBOutlet NSScrollView *scrollView;
@property (weak) IBOutlet TCGlyphView *glyphView;

- (void)calculateGlyphViewSize;

@end

@implementation TCGlyphWindowController

- (id)initWithWindow:(NSWindow *)window
{
    self = [super initWithWindow:window];
    if (self)
    {
    }
    
    return self;
}

- (void)windowDidLoad
{
    [super windowDidLoad];

    // Listen to view resize notifications
    [[[self window] contentView] setPostsFrameChangedNotifications:YES];
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
    [nc addObserverForName:NSViewFrameDidChangeNotification
                    object:[[self window] contentView]
                     queue:nil
                usingBlock:
     ^(NSNotification *note) {
         [self calculateGlyphViewSize];
     }];

    [_scrollView setHasHorizontalRuler:YES];
    [_scrollView setHasVerticalRuler:YES];
    [_scrollView setRulersVisible:YES];

    TCFontCollection *fontCollection = [(TCDocument *)[self document] fontCollection];

    // TODO: Don't just select the first font
    TCFont *font = [fontCollection fonts][0];

    TCGlyph *glyph = [[TCGlyph alloc] initWithGlyphDescription:_glyphDescription
                                               leftSideBearing:[[font hmtxTable] leftSideBearingAtIndex:[_glyphDescription glyphIndex]]
                                                  advanceWidth:[[font hmtxTable] advanceWidthAtIndex:[_glyphDescription glyphIndex]]];
    [_glyphView setGlyph:glyph];
    //[_glyphView setFont:[(TCDocument *)[self document] font]];

    [self calculateGlyphViewSize];
}

- (NSString *)windowTitleForDocumentDisplayName:(NSString *)displayName
{
    return [NSString stringWithFormat:@"%@ – %d", displayName, [_glyphDescription glyphIndex]];
}

#pragma mark - Private Methods

- (void)calculateGlyphViewSize
{
    // Calculate a space in which to place the glyph in font design units. This will
    // include a substantial blank space around the glyph.

    TCFontCollection *fontCollection = [(TCDocument *)[self document] fontCollection];

    // TODO: Don't just select the first font
    TCFont *font = [fontCollection fonts][0];

    // Visible height: head.yMax - 2 * hhea.yDescender
    NSInteger visibleBottom = 2 * [[font hheaTable] descender];
    NSInteger visibleHeight = [[font headTable] yMax] - visibleBottom;
    NSInteger visibleLeft = visibleBottom;
    NSInteger visibleWidth = visibleHeight;

    [_glyphView setBounds:NSMakeRect(visibleLeft, visibleBottom, visibleWidth, visibleHeight)];

    NSRect rect = [_scrollView bounds];

    [_glyphView setFrame:NSMakeRect(-rect.size.height, -rect.size.height, 3 * rect.size.height, 3 * rect.size.height)];

    //[[_scrollView documentView] scrollPoint:NSMakePoint(rect.size.height, rect.size.height)];

    //    [_glyphView setBounds:NSMakeRect(-665, -665, 2725, 2725)];
    [_glyphView setNeedsDisplay:YES];
}

@end

File: Typecast/TCGlyphWindowController.swift
//
//  TCGlyphWindowController.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/26/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Cocoa
import os.log
import CFF

class TCGlyphWindowController: NSWindowController {
  var glyph: TCGlyph?
  @IBOutlet weak var scrollView: NSScrollView?
  @IBOutlet weak var glyphView: TCGlyphView?

  override init(window: NSWindow?) {
    super.init(window: window)
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func windowDidLoad() {
    super.windowDidLoad()

    // Listen to view resize notifications
    self.window?.contentView?.postsFrameChangedNotifications = true
    let nc = NotificationCenter.default
    nc.addObserver(forName: .NSViewFrameDidChange,
                   object: self.window?.contentView,
                   queue: nil) { (Notification) in
      self.calculateGlyphViewSize()
    }

//    scrollView!.hasHorizontalRuler = true
//    scrollView!.hasVerticalRuler = true
//    scrollView!.rulersVisible = true

    let fontCollection = (self.document as! TCDocument).fontCollection

    // TODO: Don't just select the first font
    if let font = fontCollection?.fonts[0] {
      glyphView?.glyph = glyph
      glyphView?.font = font
      calculateGlyphViewSize()
    }
  }

  override func windowTitle(forDocumentDisplayName displayName: String) -> String {
//    if let index = glyphDescription?.glyphIndex {
//      return "\(displayName) – \(index)"
//    } else {
      return "\(displayName)"
//    }
  }

  func calculateGlyphViewSize() {

    // Calculate a space in which to place the glyph in font design units. This
    // will be limited by the xMin, yMin, xMax, and yMax values in the
    // head table.

    let fontCollection = (self.document as! TCDocument).fontCollection

    if let rect = scrollView?.bounds {
      glyphView?.frame = NSRect(x: -rect.size.width,
                                y: -rect.size.height,
                                width: rect.size.width,
                                height: rect.size.height)
    }

    // TODO: Don't just select the first font
    if let font = fontCollection?.fonts[0] {
      let head = font.headTable
      glyphView?.bounds = CGRect(x: Int(head.xMin), y: Int(head.yMin),
                                 width: Int(head.xMax) - Int(head.xMin),
                                 height: Int(head.yMax) - Int(head.yMin))
    }
    glyphView?.needsDisplay = true
  }
}
